src/language_features/document_links.rs

```diff
@@ -0,0 +1,213 @@
+use crate::editor_quote;
+use crate::position::kakoune_range_to_lsp;
+use crate::position::lsp_position_to_kakoune;
+use crate::position::lsp_range_to_kakoune;
+use crate::position::parse_kakoune_range;
+use crate::position::ranges_overlap;
+use crate::position::ranges_touch_same_line;
+use crate::EditorMeta;
+use crate::ServerId;
+
+pub fn document_links(meta: EditorMeta, ctx: &mut Context) {
+    let eligible_servers: Vec<_> = ctx
+        .servers(&meta)
+        .filter(|srv| attempt_server_capability(ctx, *srv, &meta, CAPABILITY_DOCUMENT_LINKS))
+        .collect();
+    if eligible_servers.is_empty() {
+        return;
+    }
+
+    let (first_server, _) = *eligible_servers.first().unwrap();
+    let first_server = first_server.to_owned();
+
+    let req_params = eligible_servers
+        .into_iter()
+        .map(|(server_id, _)| {
+            (
+                server_id,
+                vec![DocumentLinkParams {
+                    partial_result_params: Default::default(),
+                    text_document: TextDocumentIdentifier {
+                        uri: Url::from_file_path(&meta.buffile).unwrap(),
+                    },
+                    work_done_progress_params: Default::default(),
+                }],
+            )
+        })
+        .collect();
+    ctx.call::<DocumentLinkRequest, _>(
+        meta,
+        RequestParams::Each(req_params),
+        move |ctx, meta, results| {
+            document_links_response(meta, results, ctx);
+        },
+    );
+}
+
+fn document_links_response(
+    meta: EditorMeta,
+    results: Vec<(ServerId, Option<Vec<DocumentLink>>)>,
+    ctx: &mut Context,
+) {
+    let mut links = results
+        .into_iter()
+        .flat_map(|(server_id, v)| {
+            let v: Vec<_> = v
+                .unwrap_or_default()
+                .into_iter()
+                .map(|v| (server_id, v))
+                .collect();
+            v
+        })
+        .collect::<Vec<_>>();
+    links.sort_by_key(|(_, link)| link.range.start);
+
+    let buffile = &meta.buffile;
+    let document = match ctx.documents.get(buffile) {
+        Some(document) => document,
+        None => {
+            ctx.document_links.remove(buffile);
+            return;
+        }
+    };
+    let version = document.version;
+    let range_specs = links
+        .iter()
+        .map(|(server_id, link)| {
+            let server = ctx.server(*server_id);
+            format!(
+                "{}|LspDocumentLink",
+                lsp_range_to_kakoune(&link.range, &document.text, server.offset_encoding)
+            )
+        })
+        .join(" ");
+
+    ctx.document_links.insert(buffile.clone(), links);
+
+    let command = format!("set-option buffer lsp_document_links {version} {range_specs}");
+    let command = format!(
+        "evaluate-commands -buffer {} %§{}§",
+        editor_quote(buffile),
+        command.replace('§', "§§")
+    );
+    ctx.exec(EditorMeta::default(), command);
```

Comment:
looks correct; it's super confusing that `ctx.exec(meta, ...)` only ever uses `meta` for `meta.client`.